pkg/storage/imagestore/imagestore.go

```diff
@@ -0,0 +1,2194 @@
+func (is *ImageStore) DeleteBlob(repo string, digest godigest.Digest) error {
+	var lockLatency time.Time
+
+	if err := digest.Validate(); err != nil {
+		return err
+	}
+
+	is.Lock(&lockLatency)
+	defer is.Unlock(&lockLatency)
+
+	return is.deleteBlob(repo, digest)
+}
+
+func (is *ImageStore) deleteBlob(repo string, digest godigest.Digest) error {
+	blobPath := is.BlobPath(repo, digest)
+
+	_, err := is.storeDriver.Stat(blobPath)
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", blobPath).Msg("failed to stat blob")
+
+		return zerr.ErrBlobNotFound
+	}
+
+	// first check if this blob is not currently in use
+	if ok, _ := common.IsBlobReferenced(is, repo, digest, is.log); ok {
+		return zerr.ErrBlobReferenced
+	}
+
+	if fmt.Sprintf("%v", is.cache) != fmt.Sprintf("%v", nil) {
+		dstRecord, err := is.cache.GetBlob(digest)
+		if err != nil && !errors.Is(err, zerr.ErrCacheMiss) {
+			is.log.Error().Err(err).Str("blobPath", dstRecord).Msg("dedupe: unable to lookup blob record")
+
+			return err
+		}
+
+		// remove cache entry and move blob contents to the next candidate if there is any
+		if ok := is.cache.HasBlob(digest, blobPath); ok {
+			if err := is.cache.DeleteBlob(digest, blobPath); err != nil {
+				is.log.Error().Err(err).Str("digest", digest.String()).Str("blobPath", blobPath).
+					Msg("unable to remove blob path from cache")
+
+				return err
+			}
+		}
+
+		// if the deleted blob is one with content
+		if dstRecord == blobPath {
+			// get next candidate
+			dstRecord, err := is.cache.GetBlob(digest)
+			if err != nil && !errors.Is(err, zerr.ErrCacheMiss) {
+				is.log.Error().Err(err).Str("blobPath", dstRecord).Msg("dedupe: unable to lookup blob record")
+
+				return err
+			}
+
+			// if we have a new candidate move the blob content to it
+			if dstRecord != "" {
+				/* check to see if we need to move the content from original blob to duplicate one
+				(in case of filesystem, this should not be needed */
+				binfo, err := is.storeDriver.Stat(dstRecord)
+				if err != nil {
+					is.log.Error().Err(err).Str("path", blobPath).Msg("rebuild dedupe: failed to stat blob")
+
+					return err
+				}
+
+				if binfo.Size() == 0 {
+					if err := is.storeDriver.Move(blobPath, dstRecord); err != nil {
+						is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+						return err
+					}
+				}
+
+				return nil
+			}
+		}
+	}
+
+	if err := is.storeDriver.Delete(blobPath); err != nil {
+		is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+		return err
+	}
+
+	return nil
+}
+
+type extendedManifest struct {
+	ispec.Manifest
+
+	Digest godigest.Digest
+}
+
+func (is *ImageStore) garbageCollect(repo string) error {
+	// gc untagged manifests and signatures
+	index, err := common.GetIndex(is, repo, is.log)
+	if err != nil {
+		return err
+	}
+
+	referencedByImageIndex := []string{}
+	cosignDescriptors := []ispec.Descriptor{}
+	artifactManifests := []extendedManifest{}
+
```

Comment:
Maybe we should keep a set of digests which are either tagged, or referenced by tagged manifests, or tagged indexes, or indexes referenced by tagged indexes or indexes referenced by indexes referenced by tagged indexes and so on. 
By referenced I mean as part of a manifest list inside an index, as part of blob list, or config for a manifest, and as "subject" for either an index or a manifest.

What about cosign signatures? Even though they are tagged, we we need to check if their target digest is part of this new slice.

For notation signatures, and other generic referrers in general (maybe sboms in the future), we need to check if their subject is part of this new slice.

Note: a single pass though the index.Manifests may not be enough to identify all the relationships between digests.